numpy/core/src/multiarray/iterators.c

```diff
@@ -1772,6 +1772,9 @@ PyArray_NeighborhoodIterNew(PyArrayIterObject *x, const npy_intp *bounds,
      */
     x->contiguous = 0;
 
+    /* update the coordinates */
+    PyArray_ITER_GOTO1D(x, x->index);
```

Comment:
NIT: This is necessary when `x` was previously contiguous only, right? Maybe add that to the comment to make it clear?

Reply:
Shall I rebase the fixes or just apply a new commit?

Reply:
Either way works!  If you have large PRs, _not_ rebasing can be nice, so the reviewer can see what changed (but this is small enough it doesn't matter).

For the final merge, rebasing is nicer.  But there is now a "squash and merge" button on github and generally we started caring a bit less about the history.  So there really is no wrong answer here :).

Reply:
Done

Reply:
Maybe this trips scipy? Possibly just because there is another bug...  Scipy was always the real stress test, it just seemed unlikely to actually do things.

Reply:
That is the lead I am following right now.
If this is the reason, there is a deeper problem in scipy itself. Can someone please do the test correlation with another tool and check the results?

Reply:
I expect its not unlikely there is either a deeper problem in scipy, or some weird usage...  But, we will have to revert it in NumPy anyway if it is :(.  In the end, this is almost SciPy code put into numpy because it might be useful for someone...

@rgommers we can revert this for the sprint, you may also have to trigger the "nightly" build after that?

Reply:
>  But, we will have to revert it in NumPy anyway if it is

Unfortunately, that's how things have to work, yes

> @rgommers we can revert this for the sprint, you may also have to trigger the "nightly" build after that?

Thanks. It's not actually a nightly build, that's a misnomer - it's running on every PR.

Reply:
Yes, a misnomer, but I thought it was bi-weekly...

Reply:
No cron trigger here, just a regular job: https://github.com/scipy/scipy/blob/master/.github/workflows/linux.yml#L55

Reply:
I can't really decide if it is a bug on the original numpy implementation of PyArray_ITER_GOTO1D.
I see that scipy is creating a neighborhood iterator from a neighborhood iterator. But for sure, calling PyArray_ITER_GOTO1D inside PyArray_NeighborhoodIterNew is definitely messing with the scipy expectations. Please revert just the change to PyArray_NeighborhoodIterNew while I try figure out why this is broken.
I ran the tests here and the changes to PyArray_ITER_GOTO1D don't seem to trip scypi.

Reply:
Well, after some tests, I removed `PyArray_ITER_GOTO1D` from the constructor and called it manually outside, at the point of usage: fails the same
I used `PyArray_ITER_GOTO` instead: fails the same
By calling `_PyAIT(curneighx->_internal_iter)->coordinates[i]=0` for each coordinate, it fails, even though the original value was already `0`.
I have no clue why this would be the case, I would appreciate some help from somebody that understands what is going on on the file `./scipy/signal/correlate_nd.c.in`, function `_correlate_nd_imp`.

Reply:
Problem was on my assumption on calling PyArray_ITER_GOTO1D inside PyArray_NeighborhoodIterNew.